shared/client/src/state/witness.rs

```diff
@@ -111,11 +114,60 @@ impl WitnessStep {
         trace!("Broadcast bloom: {:?}", broadcast_bloom);
         trace!("Merkle root: 0x{}", hex::encode(broadcast_merkle.inner));
 
+        let clients_len = state.epoch_state.clients.len();
+
+        let num_actual_slices = if clients_len == 0 {
+            1 // Should not happen but just in case to avoid division by zero
+        } else {
+            clients_len.div_ceil(TRAINING_TIMES_SLICE_SIZE)
+        };
+
+        let step = state.progress.step - 1; // First step is 1, so we subtract 1 to get 0-index
+
+        // Cycle through the partitions depending on the current step, based on actual client count.
+        let slice_index = if num_actual_slices == 0 {
+            0 // Should never happen but just in case
+        } else {
+            step % (num_actual_slices as u32)
+        };
```

Comment:
is there any reason why we are calculating and sending this index in the witness? 
I think this value should be centrally calculated, tracked and updated by the coordinator in each step, and witness nodes should get it through the coordinator state

Reply:
hmm yes I can't remember why I did it like this, it's probably better to keep track of it on the coordinator I'll make the change

Reply:
ok should be changed now in the latest commit!